Rewrite dependency extraction to use mod list system

marvin-roesch opened the pull request:
As Lex stated in #4074, the dependency extraction needs a rewrite. Considering I already worked on the extraction process, anyways, I was more than happy to do this rewrite. Note, however, that it isn't completely done yet, although the only bit left to change is the location of the local mod repository (`./modrepo/` inside the run directory for testing purposes right now).

These changes were tested with some self-made JARs, not actual mods, but the rewrite doesn't change anything about mod loading order, the extraction runs as early as possible, obviously before coremod detection.

This is technically compatible with the previous mod extraction, the only difference being that modders are encouraged to provide a special field in their JAR manifest now, namely:
```
Maven-Artifact: group:name:version[:classifier]
```
It will be used to extract the JAR containing it to a mod repository and to add it to a mod list JSON. If you do not control the dependency in question, you may provide a separte file named `[dependency].jar.meta` with the same contents as the manifest would have.

The artifacts also get simple "same mod - different version" support, with the extraction candidates from mod JARs always taking priority. Any sort of version checking is both unreliable and likely to mess with modpacks (or downloaders like Curse/Twitch, to be precise), since they might want to replace a newer version of a mod with an older one.

marvin-roesch committed:
Implement dependency extraction to repository

marvin-roesch committed:
Make logging use debug level and fix gradle notation never being updated

kashike commented on src/main/java/net/minecraftforge/fml/relauncher/DependencyExtractor.java:
```diff
@@ -0,0 +1,325 @@
+package net.minecraftforge.fml.relauncher;
```
~Missing license header.~ :white_check_mark:

kashike commented on src/main/java/net/minecraftforge/fml/relauncher/DependencyExtractor.java:
```diff
@@ -0,0 +1,325 @@
+    {
+        JsonModList list = new JsonModList();
+        list.repositoryRoot = "./modrepo";
+        list.modRef = Lists.newArrayList();
```
We have Java 8 now, you can just use `new ArrayList<>();` now.

marvin-roesch committed:
Add licence header and switch to diamond operator for list initialisation

marvin-roesch committed:
Make "./libraries/" the default repository root
Remove stray log4j.properties file

mezz commented on src/main/java/net/minecraftforge/fml/relauncher/DependencyExtractor.java:
```diff
@@ -0,0 +1,344 @@
+            }
+            finally
+            {
+                if (jar != null)
```
Use `org.apache.commons.io.IOUtils#closeQuietly` instead of all this.

mezz commented on src/main/java/net/minecraftforge/fml/relauncher/DependencyExtractor.java:
```diff
@@ -0,0 +1,344 @@
+        JsonModList modList = prepareModList(modListFile);
+        File repository = new File(modList.repositoryRoot);
+        // It's a little hacky, but it should yield the best performance
+        Map<Artifact, Artifact> artifacts = modList.modRef.stream().map(Artifact::new).collect(Collectors.toMap(a -> a, a -> a));
```
Streams generally do not "yield the best performance"... If you think it's hacky just do it the regular way.

marvin-roesch replied:
The "hacky" part is the map. I'm really using it like a set with easy retrieval of existing elements. The stream performance shouldn't matter too much here, since this only happens once at startup.

kashike replied:
You can most likely use `Function::identity` instead of `a -> a`.

marvin-roesch replied:
Didn't know about it's existence, if this was Scala, I'd definitely have used `identity` xD

marvin-roesch committed:
Use IOUtils instead of unnecessary try-catch
Clarify comment about map usage

marvin-roesch committed:
Change manifest attribute name to comply with Java conventions

marvin-roesch committed:
Use Function.identity() instead of "manually" passing through inputs

marvin-roesch committed:
Add extraction test, modlist checks to come
Fix absolute files being handled incorrectly

marvin-roesch commented:
As per Lex' request, I have added a unit test for the extraction. I'll add another test for the generated modlist JSON ASAP, bedtime comes first, though. Unfortunately, the `JarFile` API doesn't allow me to generate a file in memory, so I had to resort to using a temporary directory. On the upside, the tests can be a lot more thorough this way and check the actual extraction as well.

mezz commented on src/test/java/net/minecraftforge/test/DependencyExtractionTest.java:
```diff
@@ -0,0 +1,117 @@
+     *     └─── META-INF
+     *         └    MANIFEST.MF (contains 'Maven-Artifact: net.minecraftforge.test:dependency:1.0:testing' definition)
+     */
+    private static final String JAR_BASE64 =
```
This would probably be more clear if it was just included as a jar file alongside the test.

marvin-roesch committed:
Move extractable JAR to resource

marvin-roesch committed:
Introduce versioning to modlist files, with version 2 onwards canonicalizing paths relative to the MC directory

marvin-roesch committed:
Add mod list loading test

marvin-roesch committed:
Fix dependency extractor not respecting mod list version

marvin-roesch commented:
I've added a test for the mod list loading, but it required a slight change to the modlist format, which I've put behind a new version tag that can be specified in the files. Just like the actual mod list files, the repository root now has to be marked as absolute explicitly, otherwise it will be treated as relative to the MC directory, making the system a lot more robust.

marvin-roesch committed:
Add libraries directory search

marvin-roesch committed:
Fix star imports to make @kashike shut up

modmuss50 commented:
This is all great for libs that you have control over, but most none MC forge related wont use this, how will they be handled?

gigaherz commented:
Maybe some kind of "dependency.jar.meta" file with the maven id? Can't be too hard to get [forge]gradle to generate those...

(Edit: reworded to avoid making it sound like I know what I'm saying...)

modmuss50 commented:
Yeah, there needs to be some way to specify the artifact string for libs that dont have it. Via another file might have to be best option.

marvin-roesch commented:
I definitely see the use in enabling this for arbitrary JARs, not just the ones you have control over, it's a great alternative to shading. Going with a `meta` file alongside the JAR seems reasonable. Pinging @LexManos here to get his opinion on this.

marvin-roesch committed:
Allow specifying artifact metadata externally
Cover new features in test
Use try-with-resources instead of closing streams manually

marvin-roesch committed:
Add side awareness to extraction, both for the containing JAR as well as the contained JAR

marvin-roesch commented:
I've resolved all conflicts with the stuff Lex already changed, it should functionally be on par with it. I've derped a little during rebasing, so the first commit contains some changes that technically don't belong there >.>

gigaherz commented:
Just wanted to mention, I have implemented the `Maven-Artifact` key into my libraries:

- https://github.com/gigaherz/Commons/blob/master/build.gradle#L26
- https://github.com/gigaherz/GraphLib/blob/master/build.gradle#L35

Which I include embedded into all the new 1.12 releases of my mods:

- https://github.com/gigaherz/Ender-Rift/blob/master/build.gradle#L57
- https://github.com/gigaherz/Survivalist/blob/master/build.gradle#L46
- https://github.com/gigaherz/ToolBelt/blob/master/build.gradle#L51

The current forge versions don't use the `Maven-Artifact` yet, but I'm ready for it. :D

Hope this will be useful for anyone wanting to preemptively implement this, or write documentation.

marvin-roesch committed:
Prefer the newest version of an artifact when contained in multiple JARs, still always override existing artifacts

marvin-roesch committed:
Fix tests

marvin-roesch committed:
Make extracted mods list a parent of the base one
Add removal of unneeded dependencies

marvin-roesch committed:
Make parent file reference use a relative path when possible

marvin-roesch commented:
@LexManos, pinging you here again since my implementation now slightly deviates from the initial discussion we had about this. Rather than directly writing to (any existing) `mod_list.json`, the extracted dependencies with artifact information are now listed in their own separate `extracted_mods.json` file. This file shouldn't be edited by any user (except for adding another parent in the chain), since it is wiped clean whenever the dependency configuration changes. Doing this has the added benefit that removing mods which added a dependency also causes the dependency to be removed from the mod list.

marvin-roesch committed:
Fix tests... again

marvin-roesch committed:
Don't parent the extracted mod list but add it as extra file for loading through the modlist loader

marvin-roesch committed:
Fix existing mod list not being respected for repo finding

LunNova commented:
What happens currently if two mods depend on the same coordinate but with different contents?

`[dependency].jar.meta` could be made mandatory instead of using one in the jar. This would improve performance by avoiding needing to open a jar inside a jar just to get metadata.

ForgeGradle could automatically add `[dependency].jar.meta` files using the contents of the meta-inf in the jar, or from the maven coordinates used to resolve the dependency. It could also add other data such as the build time and a hash which could be used to resolve clashes between snapshot dependencies which have the same coordinates by picking the newest one.

Here's a gradle plugin which bundles jars into mods for loading by another coremod which extracts them out. Basically the same idea as ContainedDeps but I didn't know this existed... oops: https://github.com/MinimallyCorrect/LibLoaderGradle/blob/master/src/main/java/org/minimallycorrect/libloader/LibLoaderPlugin.java

marvin-roesch commented:
I don't entirely understand what you mean by "the same coordinate" in your first paragraph. If two mods depend on something with the same Maven reference (e.g. `testgroup:testmod:1.0`), it should be reasonably safe to assume that they are in fact the same artifact. Ideally the reference is unique as-is, so if you include a dependency with a very generic name, you might just want to include the original author in the group (assuming you don't already get it directly from an actual dependency).

Making the `meta` file mandatory is unnecessary, imo. It will already be preferred upon extraction and the JAR won't be read when the `meta` file exists, so there's not really a performance issue there and it doesn't require any extra effort at build time when you depend on something that has the artifact information defined in the manifest.

I've implemented a simple system for generating `meta` files and packaging them myself as an example, it can be found here: https://github.com/PaleoCrafter/Dependency-Extraction-Example/blob/library-embedding/build.gradle#L45-L93
Ideally we'd really have it be a part of ForgeGradle, but it probably would require some extension for transitive dependencies.

LunNova commented:
`testgroup:testmod:1.0` should refer to the same thing if two mods depend on it, but `testgroup:testmod:1.0-SNAPSHOT` should be resolved to the most recent one if two mods depend on it.

> Making the meta file mandatory is unnecessary, imo. It will already be preferred upon extraction and the JAR won't be read when the meta file exists, so there's not really a performance issue there and it doesn't require any extra effort at build time when you depend on something that has the artifact information defined in the manifest.

What's the advantage of allowing it to be in the jar at the extraction stage? Ease of adding a contained dep to your mod?

marvin-roesch commented:
Ah, well, snapshots are of course a little bit of a problem. Whatever is generating the meta files would indeed need to include some timestamp information in order to resolve this properly. I don't think this information should or can in fact be determined at extraction time since we can't even rely on a standardised version format. I'll adjust my example `build.gradle` to include a timestamp for snapshots.

marvin-roesch commented:
This PR is by no means stale. Unfortunately, I ran into some issues with the snapshot thing, as retrieving the corresponding timestamp is not at all easy with Gradle's API. Also, I've found a rather critical issue with the entire system as it sits right now: Mods that are provided via the mod list are disregarded during the extraction stage. Pinging @LexManos and @cpw here in order to decide on a proper strategy for this. cpw, does your current work with modlauncher etc. potentially affect this in any way?

GGWarriorWP commented:
Is there any progress on this? I heard some modders saying that they cannot follow the forge rules if this code doesn't make it into forge... I don't know much about this, but I think if you write guidelines the people should at least have the possibility to follow them, otherwise those rules are useless

LexManos commented:
You heard wrong. Modders can always follow the guidelines no matter what goes into Forge.
The guidelines are VERY open and easily followed. This PR had some ideas to make things a little more sane, however its fallen to the way side due to the massive amount of modder resistance to imposing the standards that would be required for this.
Anyone who says they can't follow the guidelines because of *insert any reason here* is lieing to you.
The modlist json is a system that we have had implemented for years, its just nobody ever used it because it required a slight amount of work on their end.

Niko-sk2x commented:
Would br great if there was some information about *how* to use that is actually correct and not "this will work when this PR is merged" (maybe there is such information somewhere, but I wasn't able to find any). 

I've been trying to make it work for me for half a day without much of a success (partially because some gradle plugins are very stubborn in the way they do things but also because I don't exactly know how it's supposed to be done). The above mixin issue is the result of me trying to do that and failing. Even after I just did a few things manually instead of letting gradle do that.

It's especially hard when I don't have any working example. Nothing that actually does it correctly that uses mixin for instance. And even just normal coremods that do it right are hard to find. And noone I talked to knows too.

AEnterprise commented:
one mod i know that does follow the guidelines is IE, but that has the problem that whenever there is a new version (so name changes) the old ones get left behind, if the coremod also changes then old version also get loaded and things might get messy when an older and newer version of a coremod try to patch the same method

Niko-sk2x commented:
And it doesn't use mixin. The way lex specifically told me would work doesn't work (circular dependency).

JamiesWhiteShirt commented:
In order to distribute mod+core mod projects through the various launchers at present it seems you can choose only two of the following three: Following the guidelines, distributing through a single project, and having working versioning.

1. Distributing the mod and core mod as separate projects. This is compliant with the guidelines and lets the launcher handle versioning.
2. Using dependency extraction. This is compliant with the guidelines, but suffers from the versioning issue that this PR would seemingly resolve.
3. Not separating the mod and core mod. This is obviously not compliant with the guidelines, but it has every other advantage.

Therefore, it is _technically_ possible to follow the guidelines at present, but it comes with a cost. This is why some believe Forge is not truly ready to support the core mod guidelines yet, and that this PR is a necessary milestone.

I am using dependency extraction. [Here is a buildscript using it for two dependencies](https://github.com/JamiesWhiteShirt/clothesline/blob/master/build.gradle). One dependency is a core mod, the other is not even a mod. As with Immersive Engineering, old versions of the dependencies are going to stick around and require manual intervention when they are updated, which is why I am reluctant to distribute anything yet.

LexManos commented:
Starting to re-work this entire thing because its really annoying in its current state.
As stated you should be able to do whatever you want with the extracted deps as is, but Mumphry is correct, we don't have any version picking functionality.
We would need to define some form of dependency spec packed into your manifest so we can pick things.

Current work is being done here: https://github.com/LexManos/MinecraftForge/commit/6ad6c54da981e529c2253e9263436acdc00a44b6
And a *workable* solution to the issue of snapshots can be found here: https://github.com/PaleoCrafter/Dependency-Extraction-Example/pull/1/files
We'd have to get a ForgeGradle adaptation of that, but that shouldn't be hard.

marvin-roesch commented:
Closing this as Lex' version is merged. I'll be submitting another PR with unit tests for Lex' implementation.
